### PR TITLE
feat: Typescript migration

### PR DESCRIPTION
**Pull request by DavidRouyer**
### Related issues

Milestone 14

### Short description and why it's useful

- Migrate entrypoints to Typescript
- Change the Typescript configuration to allow the use of require(), dynamic imports, reflection...
- Add a few types
- Store attribute as an example for stores

### Contribution rules acceptance

- [x] I read and followed [contribution rules](https://github.com/DivanteLtd/vue-storefront/blob/master/CONTRIBUTING.md)

### CODE REVIEW

File: core/app.ts [hunk 1]
```diff
@@ -12,17 +12,20 @@ import EventBus from '@vue-storefront/core/plugins/event-bus'
...
 import registerExtensions from '@vue-storefront/core/lib/extensions'
 import i18n from '@vue-storefront/core/lib/i18n'
-import shippingMethods from '@vue-storefront/core/resource/shipping_methods.json'
 
 import store from '@vue-storefront/store'
 import coreModules from '@vue-storefront/store/modules'
 import { prepareStoreView } from '@vue-storefront/store/lib/multistore'
 
-import App from 'theme/App'
+import App from 'theme/App.vue'
```
Comment (filrak) [inline core/app.ts]: shouldn't it be resolved by default without `vue` (look at extensions in webpack config)

File: core/store/modules/attribute/types/AttributeState.ts [hunk 2]
```diff
@@ -0,0 +1,5 @@
+export default interface AttributeState {
```
Comment (filrak) [inline core/store/modules/attribute/types/AttributeState.ts]: Shouldn;t it be just 'Attribute' ?

Comment (pkarw) [reply core/store/modules/attribute/types/AttributeState.ts]: I think it should stay AttributeState as it's vuex state object

**Comment by filrak** [comment]
Regarding the types. We discussed with @pkarw that maybe it'll be better to have as few types as it is possible to allow clear understanding of data structures (eg one/two Product types, not 15 for different purposes etc). It's an open question and I don't feel like TS expert so really want to hear your opinion ;)

**Comment by pkarw** [comment]
Having store states typed has additional benefits of: intelisense support + easier to understand state structure for new comers

**Comment by DavidRouyer** [comment]
Yes, our data are in the vuex stores so it's much better if they have types!

File: core/store/modules/attribute/types/AttributeState.ts (hunk 2 above)
Comment (filrak) [reply core/store/modules/attribute/types/AttributeState.ts]: But it's also the same object as the one in View

**Comment by filrak** [comment]
I mean we should have type checking available for everything in Vuex, but not sure if we should introduce more than 1-2 types per entity

**Comment by DavidRouyer** [comment]
What do you mean by entity?

**Comment by filrak** [comment]
Product, Category, Attribute etc

**Comment by DavidRouyer** [comment]
The product model (from the Ajax call) and the product state aren't the same thing :)

**Comment by pkarw** [comment]
For me is fine. I'll merge it in Guys. Thanks @DavidRouyer great job!

**Pull request merged**